pkg/worker/events.go

```diff
@@ -0,0 +1,249 @@
+
+	for {
+		select {
+		case <-done:
+			return
+
+		case <-ticker.C:
+			stats, err := monitor.GetStatistics()
+			if err != nil {
+				return
+			}
+
+			w.eventRepo.PushContainerResourceMetricsEvent(
+				w.workerId,
+				request,
+				types.EventContainerMetricsData{
+					CPUUsed:            stats.CPU,
+					CPUTotal:           uint64(request.Cpu),
+					CPUPercent:         float32((float64(stats.CPU) * 100 / float64(request.Cpu))),
+					MemoryRSS:          stats.Memory.RSS,
+					MemoryVMS:          stats.Memory.VMS,
+					MemorySwap:         stats.Memory.Swap,
+					MemoryTotal:        uint64(request.Memory * 1024 * 1024),
+					DiskReadBytes:      stats.IO.ReadBytes,
+					DiskWriteBytes:     stats.IO.WriteBytes,
+					NetworkBytesRecv:   stats.NetIO.BytesRecv,
+					NetworkBytesSent:   stats.NetIO.BytesSent,
+					NetworkPacketsRecv: stats.NetIO.PacketsRecv,
+					NetworkPacketsSent: stats.NetIO.PacketsSent,
+					GPUMemoryUsed:      stats.GPU.MemoryUsed,
+					GPUMemoryTotal:     stats.GPU.MemoryTotal,
+					GPUType:            request.Gpu,
+				},
+			)
+		}
+	}
+}
+
+type ProcessStats struct {
+	CPU    uint64 // in millicores
+	Memory process.MemoryInfoStat
+	IO     process.IOCountersStat
+	NetIO  net.IOCountersStat
+	GPU    GPUInfoStat
+}
+
+type GPUInfoStat struct {
+	MemoryUsed  uint64
+	MemoryTotal uint64
+}
+
+type ProcessMonitor struct {
+	pid       int32
+	devices   []specs.LinuxDeviceCgroup
+	lastIO    process.IOCountersStat
+	lastNetIO net.IOCountersStat
+}
+
+func NewProcessMonitor(pid int, devices []specs.LinuxDeviceCgroup) *ProcessMonitor {
+	return &ProcessMonitor{pid: int32(pid), devices: devices}
+}
+
+func (m *ProcessMonitor) GetStatistics() (*ProcessStats, error) {
+	processes, err := m.findProcesses()
+	if err != nil {
+		return nil, err
+	}
+
+	gpuStat := m.fetchGPUMemory()
+
+	netIOStat, err := m.fetchNetworkIO()
+	if err != nil {
+		return nil, err
+	}
+
+	ioStat, err := m.fetchIO(processes)
+	if err != nil {
+		return nil, err
+	}
+
+	millicores := m.fetchCPU(processes)
+	memoryStat := m.fetchMemory(processes)
+
+	return &ProcessStats{
+		CPU:    uint64(millicores),
+		Memory: *memoryStat,
+		IO:     *ioStat,
+		NetIO:  *netIOStat,
+		GPU:    *gpuStat,
+	}, nil
+}
+
+func (m *ProcessMonitor) fetchGPUMemory() *GPUInfoStat {
+	stat := &GPUInfoStat{}
+	for _, device := range m.devices {
+		if device.Minor == nil || !device.Allow {
+			continue
+		}
+		stats, err := GetGpuMemoryUsage(int(*device.Minor))
+		if err == nil {
+			stat.MemoryUsed += uint64(stats.UsedCapacity)
+			stat.MemoryTotal += uint64(stats.TotalCapacity)
+		}
+	}
+	return stat
+}
+
+// fetchNetworkIO gets network IO for all NICs globally, regardless of process.
+// Avoid using this data for external purposes.
+// TODO: Look into per-process network IO
+func (m *ProcessMonitor) fetchNetworkIO() (*net.IOCountersStat, error) {
+	counters, err := net.IOCountersByFile(false, filepath.Join("/proc", fmt.Sprint(m.pid), "net/dev"))
```

Comment:
Interesting. Wouldn't this just get the network io for the process with `m.pid` so just the worker?

I'm confused how this file includes network io for all processes

Reply:
Have you tried `findChildProcesses` and summing up the io of all the container processes?

Reply:
Your thoughts are exactly right. I could just call the function IOCounters() instead if it clearer. 

I found that /proc/{pid}/nev/dev gives me the exact results as /proc/net/dev. So I did this just to show that it exists, but doesn't really matter.

Reply:
I have - processes don't have a Net IO Counter function. Notice I'm using the `net` package here and this function in particular lets me point to a proc path.

Reply:
interesting. Is it because maybe the worker is pid 1, so it already gets the sum of all the child io?

Reply:
ah nvm, just tested it myself. its all the same it seems

Reply:
I'm just going to switch to `counters, err := net.IOCounters(false)` to avoid the confusion and add a couple of comments about this.